### PR TITLE
Add support for Spark 3.0

### LINKED ISSUE #1412: [Feature] Spark3.0 support
### Feature description

Spark3 is currently in RC. Will there be support for Spark3 in the next release version (v8) or will we have to wait for v9 ? 
more precisely, do you guys plan to start supporting spark3 only when official spark3 release is made ?
- spark3 now support only scala 2.12

### PR DESCRIPTION
**Pull request by jbaiera**
This PR adds support for Spark version 3.0. 

Closes #1412

Reviewer's note: The PR is best reviewed by evaluating the changes on each commit, since 6c0b177 is just a source copy operation from the Spark 2.x project.

### CODE REVIEW

**Comment by jbaiera** [comment]
Known issue with Spark 3.0 #1593

**Comment by lyogev** [comment]
@jbaiera I was able to build your branch, with JDK 11 although I had to export ```JAVA8_HOME``` to my JDK8 dir.
Tested it with a simple writing test and it seems to work well, will update after further testing.

File: spark/sql-30/src/itest/java/org/elasticsearch/spark/integration/SparkSQLSuite.java [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+package org.elasticsearch.spark.integration;
+
+import org.elasticsearch.hadoop.fixtures.LocalEs;
+import org.junit.ClassRule;
+import org.junit.rules.ExternalResource;
+import org.junit.runner.RunWith;
+import org.junit.runners.Suite;
+
+@RunWith(Suite.class)
+@Suite.SuiteClasses({ AbstractJavaEsSparkSQLTest.class })
```
Comment (aimiebell) [inline spark/sql-30/src/itest/java/org/elasticsearch/spark/integration/SparkSQLSuite.java:28]: I detect that this code is problematic. According to the [Bad practice (BAD_PRACTICE)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#bad-practice-bad-practice), [Nm: Method names should start with a lower case letter (NM_METHOD_NAMING_CONVENTION)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#nm-method-names-should-start-with-a-lower-case-letter-nm-method-naming-convention).
Methods should be verbs, in mixed case with the first letter lowercase, with the first letter of each internal word capitalized.

**Comment by paulrossmeier** [comment]
@jbaiera I was able to build you branch as well - using JDK 11 although I had to export JAVA8_HOME to my JDK8 dir.

I have tested this writing a few tables to elastic - one with ~ 9 million rows - using "WAN" setting as I am indexing into a hosted cluster.

I have had no issues this far

**Review by danhermann — APPROVED** [review]
LGTM. I've walked through the code with @jbaiera and done some basic testing both reading from and writing to ES in Spark3.

**Pull request merged**